design/one-pager-merge-jets-proposal.md

```diff
@@ -0,0 +1,402 @@
+a generic controller that is based on crossplane-runtime generic managed
+reconciler. The schema of the CRDs are derived from Terraform providers and the
+generic controller calls Terraform under the hood. With Terrajet, we're able to
+generate Crossplane providers that are on par with Terraform providers in terms
+of coverage in the matter of days. The API of the generated resources are fully
+compliant with [Crossplane Resource Model (XRM)][xrm-doc]. You can read more
+about its internals in [the deep dive blog post series][deep-dive-blogs].
+
+Before releasing the first Terrajet-based providers, we have had a discussion
+whether we'd like to have them as separate providers or have the
+Terrajet-generated CRDs added to the existing classic providers in the [provider
+strategy doc][strategy-doc]. The decision for initial releases of the first
+three Jet-based providers ([AWS][jet-aws], [Azure][jet-azure] and
+[GCP][jet-gcp]) was to have them as separate providers for the reasons stated
+[here](https://github.com/crossplane/crossplane/blob/master/design/design-doc-provider-strategy.md#decision-for-initial-releases).
+
+We called out that we'll revisit this decision after more usage and community
+feedback, which was lacking most probably because we were asking people (in
+Slack and in https://github.com/crossplane/crossplane/pull/2701) for their
+opinions about providers that they have not used yet. Today we have more data
+and more awareness in the community, and having two different implementations
+has been confusing for a lot of users, so a revisit of the decision regarding
+AWS, Azure and GCP is due.
+
+## Goals
+
+We have the following goals for _the big three_ providers:
+
+* The cloud vendors should maintain their Crossplane provider.
+* The users should have the ability to provision all of their infrastructure
+  through Crossplane, i.e. coverage.
+* The providers should be mature enough for users to depend on them in
+  production, i.e. maturity.
+
+Today, we have two Crossplane providers for each of the big three clouds.
+* AWS
+  * Classic with [124 CRDs][provider-aws-crds] with 37 beta CRDs.
+  * Jet with [763 CRDs][jet-aws-crds], no beta CRDs.
+* Azure
+  * Classic with [21 CRDs][provider-azure-crds] with 7 beta CRDs.
+  * Jet with [647 CRDs][jet-azure-crds], no beta CRDs.
+* Google Cloud
+  * Classic with [27 CRDs][provider-gcp-crds] with 11 beta CRDs.
+  * Jet with [438 CRDs][jet-gcp-crds], no beta CRDs.
+
+Both variants of each reach those goals at different levels but none of them are
+maintained by their cloud vendor. The aim of this document is to capture a
+decision about the future of these provider implementations with the goals
+stated above in mind.
+
+## Proposal
+
+There were three options in the provider strategy doc:
+* Option A: Completely separate providers.
+* Option B: One provider but Jet-based resources in their own API group.
+* Option C: One provider and no Jet vs classic difference. Only API version
```

Comment:
it's not immediately clear to me what happens in option C here when there are conflicting resources between classic and jet.  i.e. if provider-aws already has `foo.aws.crossplane.io/v1beta1` and jet has `foo.jet.aws.crossplane.io/v1alpha2`, what happens when they get merged? is there only 1 that wins since they are merged into the same provider and api group?

maybe this is describe later in the doc, but it's the very first question that came to my mind and should perhaps be addressed more up front.

Reply:
This is even worse in Azure, mainly AKS implementation, which in classic is very basic and missing many features. If the deciding criteria will be: use jet if classic resource does not exist functionality will be degraded. This implies that for merging via option C each resource has to be crossed check about functionality and this will require substantial work.

Reply:
@jbw976 I've added a list of tangible steps about how it's going to be, let me know if that doesn't cover your concern. The idea is that whatever exists already will stay as is and new CRDs will be added with Terrajet without the `jet` grouping, just like adding another CRD using ACK.

@andrzej-natzka Initially, no such conversion will take place. But there are cases where we might want to do Jet->classic or classic->Jet and I dropped a guideline about those. @negz has suggested we do classic->Jet for all existing ones at some point by setting a goal [here](https://github.com/crossplane/crossplane/pull/2930#discussion_r817263352), you can follow that discussion there and add your thoughts.